lib/NormalModule.js

```diff
@@ -69,260 +40,385 @@ function contextify(options, request) {
+			emitFile: (name, content, sourceMap) => {
+				this.assets[name] = this.createSourceForAsset(name, content, sourceMap);
+			},
+			options: options,
+			webpack: true,
+			sourceMap: !!this.useSourceMap,
+			_module: this,
+			_compilation: compilation,
+			_compiler: compilation.compiler,
+			fs: fs,
+		};
+
+		compilation.applyPlugins("normal-module-loader", loaderContext, this);
+		if(options.loader)
+			Object.assign(loaderContext, options.loader);
+
+		return loaderContext;
+	}
+
+	createSource(source, resourceBuffer, sourceMap) {
+		// if there is no identifier return raw source
+		if(!this.identifier) {
+			return new RawSource(source);
 		}
 
-		var resourceBuffer = result.resourceBuffer;
-		var source = result.result[0];
-		var sourceMap = result.result[1];
+		// from here on we assume we have an identifier
+		const identifier = this.identifier();
+
+		if(this.lineToLine && resourceBuffer) {
+			return new LineToLineMappedSource(
+				source, identifier, asString(resourceBuffer));
+		}
 
-		if(!Buffer.isBuffer(source) && typeof source !== "string") {
-			return callback(module.error = new ModuleBuildError(module, new Error("Final loader didn't return a Buffer or String")));
+		if(this.useSourceMap && sourceMap) {
+			return new SourceMapSource(source, identifier, sourceMap);
 		}
-		source = asString(source);
-		if(module.identifier && module.lineToLine && resourceBuffer) {
-			module._source = new LineToLineMappedSource(source, module.identifier(),
-				asString(resourceBuffer));
-		} else if(module.identifier && module.useSourceMap && sourceMap) {
-			module._source = new SourceMapSource(source, module.identifier(), sourceMap);
-		} else if(module.identifier) {
-			module._source = new OriginalSource(source, module.identifier());
+
+		return new OriginalSource(source, identifier);
+	}
+
+	doBuild(options, compilation, resolver, fs, callback) {
+		this.cacheable = false;
+		const loaderContext = this.createLoaderContext(resolver, options, compilation, fs);
+
+		runLoaders({
+			resource: this.resource,
+			loaders: this.loaders,
+			context: loaderContext,
+			readResource: fs.readFile.bind(fs)
+		}, (err, result) => {
+			if(result) {
+				this.cacheable = result.cacheable;
+				this.fileDependencies = result.fileDependencies;
+				this.contextDependencies = result.contextDependencies;
+			}
+
+			if(err) {
+				return callback(this.error = new ModuleBuildError(this, err));
+			}
+
+			const resourceBuffer = result.resourceBuffer;
+			const source = result.result[0];
+			const sourceMap = result.result[1];
+
+			if(!Buffer.isBuffer(source) && typeof source !== "string") {
+				this.error = new ModuleBuildError(this, new Error("Final loader didn't return a Buffer or String"));
+				return callback(this.error);
+			}
+
+			this._source = this.createSource(asString(source), resourceBuffer, sourceMap);
+			return callback();
+		});
+	}
+
+	disconnect() {
+		this.built = false;
+		super.disconnect();
+	}
+
+	markModuleAsErrored() {
```

Comment:
Could you rewrite this method to take an `error` argument and be the only place where `this.error = error` is set? This should reduce the number of places where `this.error` is written and make the control flow more clear.